jsconfuser: new plugin

echo094 opened the pull request:
## What this adds

A new `-t jsconfuser` target that reverses [JS-Confuser](https://github.com/MichaelXF/js-confuser)
output, built against pinned encoder commit `31c5a47` (js-confuser 2.1.3). It joins the
existing `common` / `jjencode` / `sojson` / `sojsonv7` / `obfuscator` targets and is reachable
as `npm run dejsc`.

The plugin is a flat sequence of independent `traverse` passes, each reversing one encoder
transform by AST shape. Matching is structural throughout — never by identifier name, because
the encoder's `RenameVariables` stage (Order 30) reassigns every identifier including its own
runtime helpers', which silently defeats any name-keyed matcher. Three passes construct an
`isolated-vm` isolate to evaluate the encoder's own runtime helper where a value has to be
computed rather than read.

**Coverage:** every one of the encoder's 23 `Order` transforms either has a decode pass or a
recorded reason it needs none. `docs/jsconfuser.md` states what is reversed, what is not, and
why — `RenameVariables`, `RenameLabels` and `ObjectExtraction` are irreversible by design
(the information is destroyed on the encode side, not merely hidden), and the doc says so
rather than leaving them looking unimplemented.

## What this changes outside the new target

- **Two shared visitors extended**, both additive: `calculate-constant-exp.js` gains a
  `LogicalExpression` fold, and `prune-if-branch.js` now splices the surviving branch instead
  of planting the whole `BlockStatement`. Existing targets exercise both and are unaffected —
  see the test plan.
- **`vitest.config.js` gains an explicit `testTimeout`.** One fixture landed past vitest's 5s
  default on CI while passing locally; the runner is several times slower per case and that
  fixture was the only one whose CI cost crossed the line.

## Review notes

- **Commit granularity is one commit per unit** — 37 `visitor/*`, 11 `plugin/jsconfuser`,
  4 `utility/*`, 5 package-wide `decode-js`. Reviewing commit-by-commit should read cleanly;
  each was linted and suite-tested as it landed, not only at the tip.
- **Four orderings in the history are load-bearing** and are not arbitrary: `rgf.js` and
  `plugin/jsconfuser.js` are genuinely circular (decoding an eval-wrapped sub-program re-runs
  the whole pipeline) so they share one commit; `control-flow-graph.js` depends on
  `control-flow.js`; `function-length.js` and `dispatcher.js` both depend on
  `variable-masking.js`; and two control-flow tests read fixtures introduced by an earlier
  commit.
- **Two fixtures carry a scope that does not match their filename**, deliberately:
  `string-splitting` has no decode pass of its own so its fixture is scoped
  `plugin/jsconfuser`, and the `control-flow-flattening` fixture is named for the encoder
  transform while the pass it exercises is `control-flow-graph.js`.

## Test plan

Run at the tip of this branch.

| check | result |
|---|---|
| `npx vitest run` | 37 files, 45 suites, 309 tests — all pass |
| `npm ci && npx vitest run` (the CI pair, `npm_config_build_from_source=true`) | same, all pass |
| `npx eslint src/` | clean — a zero baseline, not a tolerated-file list |
| full-corpus runtime equivalence | 117 / 117 match, 0 vacuous |
| decoder-introduced unbound references | 0 of 117 |
| decoded ÷ pre-obfuscation source | 1.34× |
| decoded ÷ obfuscated input | 0.001× |

**What the corpus run is and is not.** 117 samples built from the 39 committed fixture
sources (24 primary + 15 under `test/jsconfuser/rename-variables/`), 3 independent `high`
encodes each. Its *sources are all committed*, so it covers no program the fixtures don't —
what it buys is **encoder-randomness breadth**, which is the only thing that exercises a
probability-gated transform (Flatten fires in roughly half of `high` runs). Only the
obfuscated bytes are untracked, and they regenerate on demand though never byte-for-byte,
since the encoder samples randomly. The durable gate is the committed fixture triples, which
is what CI enforces.

**The unbound-reference control is worth stating**, because it is what makes the 0 meaningful:
running the same census over the `.obf.js` *inputs* reports one unbound reference (`x`, on
`control-flow-flattening-minify.2`), so that one is the encoder's and not something this
decoder introduced.

**Runtime correctness alone proves little here and is not offered as the main result.** Every
pass fails closed — a pass that cannot recognise its shape leaves the input untouched — so a
completely broken decoder and a working one produce identical *runtime behaviour* and differ
only in output *shape*. That is why the size ratios and the unbound-reference census are in
the table alongside the pass count.

einstein95 commented on src/visitor/jsconfuser/stack.js:
```diff
@@ -0,0 +1,356 @@
+  while (changed) {
+    checkStackInvalid(path, invalid)
+    if (!checkChangeValid(invalid, used)) {
+      path.replaceWith(prase(orig_code))
```
`ReferenceError: prase is not defined`

einstein95 commented:
```
/Users/pc/decode-js/src/visitor/jsconfuser/stack.js:197
  const stk_name = path.node.params[0].argument.name
                                   ^

TypeError: Cannot read properties of undefined (reading '0')
    at tryStackReplace (/Users/pc/decode-js/src/visitor/jsconfuser/stack.js:197:36)
    at processStackParam (/Users/pc/decode-js/src/visitor/jsconfuser/stack.js:298:15)
    at Identifier (/Users/pc/decode-js/src/visitor/jsconfuser/stack.js:324:9)
    at NodePath._call (/Users/pc/decode-js/node_modules/@babel/traverse/lib/path/context.js:49:20)
    at NodePath.call (/Users/pc/decode-js/node_modules/@babel/traverse/lib/path/context.js:39:18)
    at NodePath.visit (/Users/pc/decode-js/node_modules/@babel/traverse/lib/path/context.js:85:31)
    at TraversalContext.visitQueue (/Users/pc/decode-js/node_modules/@babel/traverse/lib/context.js:89:16)
    at TraversalContext.visitSingle (/Users/pc/decode-js/node_modules/@babel/traverse/lib/context.js:65:19)
    at TraversalContext.visit (/Users/pc/decode-js/node_modules/@babel/traverse/lib/context.js:112:19)
    at traverseNode (/Users/pc/decode-js/node_modules/@babel/traverse/lib/traverse-node.js:22:17)
```
[code.js.zip](https://github.com/user-attachments/files/18446329/code.js.zip)

echo094 committed:
chore(decode-js): relicense under GPLv3

Replace the MIT text with the official GNU GPL v3 document verbatim as fetched
from gnu.org, and declare it in package.json, which carried no license field at
all until now.

GPL-3.0-or-later rather than -only, following the GNU recommendation that the
default grant allow any later version.

Prior contributions were made under MIT, which is GPL-compatible, so the
combined work may be distributed under GPLv3. The verbatim license document has
no place for a project copyright line; the pre-existing "Copyright (c) 2022
echo094" attribution wants a home in the README when that file is rewritten.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(utility/logger): add a debug logger gated by one switch

Per-pass progress tracing was scattered across ad-hoc console calls that could
not be turned off. Route it through one module so a decode run is silent by
default and verbose only when -v or DECODE_JS_DEBUG=1 asks for it.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(utility/binding-def): resolve a binding to its real definition site

`binding.path` names the definition only for the plainest declaration spelling.
A split declaration and a parameter-hoisted name both leave the binding intact
while moving the value elsewhere, so the rename-proof "binding whose path is a
FunctionDeclaration" idiom this decoder uses throughout fails closed on both.
Resolve through the binding's writes instead, and return null rather than guess
when there is more than one.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(utility/check-func): add a subsequence check for generated code

Matching a runtime helper by its exact source text breaks under whitespace and
spelling differences between encoder versions. Test for the pattern as an
ordered subsequence instead, which tolerates the interleaving without accepting
an unrelated function.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(utility/safe-func): add the fail-closed AST edit helpers

Deleting a node, reading a literal, resolving a name and replacing a value all
have to survive a scope re-crawl that can invalidate the very binding they were
given. Collect them behind one module that declines instead of throwing, so a
pass that cannot complete an edit leaves the tree untouched.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/calculate-constant-exp): short-circuit constant logical expressions

A constant test on the left of && or || is a branch that never varies, and
leaving it in place keeps the dead side alive for every later pass to walk
around. Fold it here, and un-compute a folded string that lands in a member key
so the key reaches downstream matchers in one spelling rather than two.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/string-concealing): decode js-confuser's concealed string arrays

The per-block retrieval wrapper has the same shape whatever string encoding is
in play, default base91 or a customStringEncodings implementation, so match it
structurally and run the encoder's own decode function in an isolate rather than
reimplementing the encoding here.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/string-compression): decode js-confuser 1.x compressed string blobs

stringCompression was a 1.1.2 option that did not survive the 2.0.0 Babel
rewrite, so this targets 1.x output only. The pinned encoder cannot produce a
sample, which is why the pass carries no fixtures.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/global-concealing): restore js-confuser's concealed globals

Every case of the generated switch is the same single return of a computed
member access, real names and decoy padding alike, so the key to name map is
recoverable from shape alone and is enough to rewrite every call site.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/duplicate-literal): restore js-confuser's deduplicated literals

The array is matched all-or-nothing, so the spellings Minify leaves behind have
to be recognized here rather than deferred to the generic constant folding later
in the pipeline: one unrecognized element fails the whole array closed and
strands every reference to it.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/calculator): resolve js-confuser's arithmetic dispatch calls

Parameters are matched positionally rather than by name: the encoder's own
source names are mangled by RenameVariables long before a decoder sees the
output.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/dead-code): remove js-confuser's dead-code guards

DeadCode's guard is matched and removed as one self-contained unit rather than
folded through the shared constant-folding passes, so its dummy-predicate test
stays independent of the one opaque-predicates uses.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/opaque-predicates): resolve js-confuser's opaque predicates

An 'in' check reads false exactly while nothing adds the property, so soundness
comes from the anchor's reference set rather than from its declared shape. The
shape was only ever a proxy, and it is one this decoder's own control-flow pass
rewrites.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/variable-masking): un-mask js-confuser's stack-slot variables

Slots resolve through bindings rather than through the stack parameter's name
text. renameVariables reuses one name across a shadowing outer/inner pair, and a
name-keyed match then reads a nested function's stack as the enclosing one's.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/function-length): unwrap js-confuser's preserveFunctionLength stubs

The template's descriptor keys arrive quoted, the form Preparation normalizes
every object key to, or plain when Minify converts the valid-identifier ones
back. Read either shape rather than pinning one.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/dispatcher): restore js-confuser's dispatched function tables

Entries reach the matcher already masked, and the fns entry is the only place
an exact parameter count can be read, so the un-masking is driven from there
rather than inferred from the call sites.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/control-flow): resolve js-confuser's entangled control-flow literals

Recovers the numbers, booleans and XOR-ed strings that ControlFlowFlattening
entangles into its state vector. The graph walk that consumes these primitives
lands separately.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/control-flow-graph): rebuild js-confuser's flattened control flow

Walks the transition graph into a block DAG and folds the branch nodes back
into real if/else. Every runtime-helper lookup resolves through binding
structure rather than the name suffixes the encoder gives them, which
RenameVariables scrambles wholesale while necessarily preserving the bindings.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/flatten): inline js-confuser's Flatten scope objects

Every property of the flat object proxies one outer-scope identifier in one of
three exact shapes. A property matching none of them fails the whole object,
which is treated as 'not this pattern' rather than decoded halfway.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
fix(visitor/prune-if-branch): splice a pruned branch instead of leaving its block

A bare block is transparent to the language but not to a matcher: deFlatten
reads a wrapper's last statement and requires a return of a call, so a pruned if
left behind as a block failed that match and stranded an entire scope-object
layer undecoded. Splice only where it is sound, in a statement list with nothing
block-scoped to relocate.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/moved-declarations): reverse js-confuser's parameter packing

MovedDeclarations retypes a nested declaration to an expression, appends its
name to the enclosing parameter list and prepends a guarded re-assignment. The
guard exists because the slot is a real parameter a caller could supply, so
reversing it means removing the slot as well as the guard.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/ast-scrambler): undo js-confuser's statement merging

The no-op helper injected once per Program is matched structurally: its name is
a random placeholder and carries nothing to match on.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/integrity): remove js-confuser's integrity checks

The memoized hash is stored on a plain, non-computed member of the wrapper's
own name, which is what identifies the cache slot to remove without depending on
the property's random spelling.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/lock): remove js-confuser's date, domain and tamper locks

Each lock template is matched by the shape of its test closure and its
countermeasure call, so an unrecognized variant leaves the guard in place rather
than silently disarming a check the decoder did not understand.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/pack): unwrap js-confuser's packed payload

The payload is a Function(<string>, <string>) call whose body is re-parsed and
spliced back into the program. The trailing return the wrapper adds is reversed
too, so the decoded result is still a valid script rather than an expression.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(visitor/anti-tooling): restore shapes js-confuser 1.x anti-tooling strips

antiTooling was a 1.1.8 option that did not survive the 2.0.0 Babel rewrite, so
this targets 1.x output only and its fixtures are hand-written rather than
encoder-generated.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(plugin/jsconfuser): assemble the jsconfuser decode pipeline

Schedules the passes in reverse encoder-stage order, with the second visits a
later stage's residue forces: the string layer and VariableMasking both run
again after the control-flow decode, since flattening rewrites reference sites
that only become resolvable once its own graph is.

rgf.js lands here rather than with the other visitors because the two are
mutually dependent - decoding an eval-wrapped sub-program means running this
whole pipeline over it, so rgf.js imports the plugin that imports rgf.js.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/rgf): cover the eval-wrapper decode

Held back from the pass's own commit because the pass could not land without
the plugin it recurses through, and the plugin commit should not also carry a
test suite.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
feat(decode-js): register the jsconfuser target on the CLI

Adds the target to the -t roster, so the new decoder is reachable the same way
every other target is.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
chore(decode-js): add a dejsc npm script

Every other target has a named script; the new one should not be the exception
that has to be invoked through main.js by hand.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(decode-js): raise the vitest timeout for whole-pipeline fixtures

The jsconfuser fixtures decode real encoder output end to end, and the heaviest
sits near 2s on a fast dev machine - inside vitest's 5s default, but with too
little margin for a shared CI runner, where it has already timed out while
passing locally. A genuine hang still fails, just later.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the control-flow-flattening decodes

Nine triples covering the entry-harness shapes the graph decoder has to
recognize - Function-level, outlined nested function, bare Program-level - and
the later encoder stages that rewrite them underneath: minify, moved
declarations, the AST scrambler, a throw terminal and a dispatcher.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the string-layer combinations

The string layer is reference-count-gated and swept across stage boundaries, so
what it leaves behind depends on what else consumed its output. Each triple
pairs it with a different consumer.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the calculator combinations

One pre-CFF and one post-CFF, because the calculator's dispatch calls are
reachable at two different points in the pipeline and only the second needs the
assigned-holder spelling.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the rgf combinations

RGF re-enters the whole pipeline on its sub-program, so each of these exercises
the plugin twice over - once outside the eval wrapper and once within it.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the dead-code and opaque-predicates decode

Both passes key on the same dummy-function anchor from opposite directions, and
this triple is what keeps one from removing the other's evidence.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the flatten and function-length decode

preserveFunctionLength wraps the Flatten wrapper, so the two reversals have to
compose in one direction only.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the finalizer and pack decodes

Both run at the end of the encoder's own order, over output every other
transform has already produced, which is why they get their own triples rather
than riding along with an earlier stage's.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the masking trio and the template/regex case

The trio is the one combination where CFF, the dispatcher and variable masking
all feed each other. The template/regex case covers syntax the corpus builder
cannot generate.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/control-flow-graph): pin the inline-flattened collapse

These two read their samples out of test/jsconfuser/ rather than carrying their
own, so they cannot land before the fixture that provides one.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/flatten): pin the flatten decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/lock): pin the lock decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/rgf): pin the rgf decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/dispatcher): pin the dispatcher decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/dead-code): pin the dead-code decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/calculator): pin the calculator decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/global-concealing): pin the global-concealing decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/opaque-predicates): pin the opaque-predicates decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(plugin/jsconfuser): pin the string-splitting decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Scoped to the plugin rather than to a visitor: StringSplitting has no decoder
pass of its own, so this fixture pins the pipeline's behaviour, not one file's.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/string-concealing): pin the string-concealing decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/variable-masking): pin the variable-masking decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/duplicate-literal): pin the duplicate-literal decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/moved-declarations): pin the moved-declarations decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/ast-scrambler): pin the ast-scrambler decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
test(visitor/control-flow-graph): pin the control-flow-flattening decode under renameVariables

RenameVariables runs at encoder order 30, over almost every transform this decoder
reverses, and it invalidates a name-keyed matcher silently and totally - the pass
declines on every application at once, with correct runtime output. This fixture
either pins a real bug that uncovered, or stands as proof the pass was cleared.

Named for the encoder transform rather than the decoder file, matching the other
fixtures here; the pass it exercises is control-flow-graph.js.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
docs(plugin/jsconfuser): document what the jsconfuser target reverses

Per-target detail belongs off the entry page so the README can stay a roster.
This says what is reversed, what is not, and why - the asymmetry with the
obfuscator target's inline feature list is deliberate.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>

echo094 committed:
docs(decode-js): rewrite the README in English and lead with a target table

The entry page is a roster of supported targets, and it is where the project
copyright and the GPLv3 pointer live - a verbatim GNU LICENSE has nowhere to put
either.

Signed-off-by: echo094 <20028238+echo094@users.noreply.github.com>
